Userland/Libraries/LibJS/Runtime/ArrayPrototype.cpp

```diff
@@ -1551,42 +1551,62 @@ ThrowCompletionOr<void> array_merge_sort(GlobalObject& global_object, FunctionOb
+    // 1. If comparefn is not undefined and IsCallable(comparefn) is false, throw a TypeError exception.
+    auto comparefn = vm.argument(0);
+    if (!comparefn.is_undefined() && !comparefn.is_function())
+        return vm.throw_completion<TypeError>(global_object, ErrorType::NotAFunction, comparefn.to_string_without_side_effects());
 
+    // 2. Let obj be ? ToObject(this value).
     auto* object = TRY(vm.this_value(global_object).to_object(global_object));
 
+    // 3. Let len be ? LengthOfArrayLike(obj).
     auto length = TRY(length_of_array_like(global_object, *object));
 
-    MarkedVector<Value> items(vm.heap());
-    for (size_t k = 0; k < length; ++k) {
-        auto k_present = TRY(object->has_property(k));
+    // 4. Let SortCompare be a new Abstract Closure with parameters (x, y) that captures comparefn and performs the following steps when called:
+    Function<ThrowCompletionOr<double>(Value, Value)> sort_compare = [&](auto x, auto y) -> ThrowCompletionOr<double> {
+        // a. Return ? CompareArrayElements(x, y, comparefn).
+        return TRY(compare_array_elements(global_object, x, y, comparefn.is_undefined() ? nullptr : &comparefn.as_function()));
+    };
 
-        if (k_present) {
-            auto k_value = TRY(object->get(k));
-            items.append(k_value);
-        }
-    }
+    // 6. Let sortedList be ? SortIndexedProperties(obj, len, SortCompare, true).
+    auto sorted_list = TRY(sort_indexed_properties(global_object, *object, length, sort_compare, true));
+
+    // 7. Let itemCount be the number of elements in sortedList.
+    auto item_count = sorted_list.size();
 
-    // Perform sorting by merge sort. This isn't as efficient compared to quick sort, but
-    // quicksort can't be used in all cases because the spec requires Array.prototype.sort()
-    // to be stable. FIXME: when initially scanning through the array, maintain a flag
-    // for if an unstable sort would be indistinguishable from a stable sort (such as just
-    // just strings or numbers), and in that case use quick sort instead for better performance.
-    TRY(array_merge_sort(global_object, callback.is_undefined() ? nullptr : &callback.as_function(), items));
+    // 8. Let j be 0.
+    size_t j = 0;
 
-    for (size_t j = 0; j < items.size(); ++j)
-        TRY(object->set(j, items[j], Object::ShouldThrowExceptions::Yes));
+    // 9. Repeat, while j < itemCount,
+    for (; j < item_count; ++j) {
+        // a. Perform ? CreateDataPropertyOrThrow(obj, ! ToString(𝔽(j)), sortedList[j]).
 
-    // The empty parts of the array are always sorted to the end, regardless of the
-    // compare function. FIXME: For performance, a similar process could be used
-    // for undefined, which are sorted to right before the empty values.
-    for (size_t j = items.size(); j < length; ++j)
+        // FIXME: Spec issue: The above step should likely be:
```

Comment:
Are you filing an issue for this? :)

Reply:
https://github.com/tc39/proposal-change-array-by-copy/issues/97

Reply:
I'll link that in the comment here for reference too

Reply:
Nice, ty!